lib/active_record/pg_extensions/pessimistic_migrations.rb

```diff
@@ -21,6 +21,30 @@ def change_column_null(table, column, nullness, default = nil)
         end
         super
       end
+
+      # several improvements:
+      #  * support if_not_exists
+      #  * delay_validation automatically creates the FK as NOT VALID, and then immediately validates it
+      #  * if delay_validation is used, and the index already exists but is NOT VALID, it just re-tries
+      #    the validation, instead of failing
+      def add_foreign_key(from_table, to_table, delay_validation: false, if_not_exists: false, **options)
+        # pointless if we're in a transaction
+        delay_validation = false if open_transactions.positive?
+        options[:validate] = false if delay_validation
+
+        options = foreign_key_options(from_table, to_table, options)
+
+        if if_not_exists || delay_validation
+          scope = quoted_scope(options[:name])
+          valid = select_value(<<~SQL, "SCHEMA")
+            SELECT convalidated FROM pg_constraint INNER JOIN pg_namespace ON pg_namespace.oid=connamespace WHERE conname=#{scope[:name]} AND nspname=#{scope[:schema]}
+          SQL
+          return if valid == true && if_not_exists
+        end
+
+        super(from_table, to_table, **options) unless valid == false
```

Comment:
If if_not_exists and delay_validation are both false, will valid be defined here?

Reply:
it will be nil (even though the conditional above didn't execute, ruby will materialize the local variable with a default of nil)